src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java

```diff
@@ -357,12 +404,20 @@ protected boolean inBounds(DynamicObject matchData, int index) {
         }
     }
 
+    // because the factory is not constant
+    @TruffleBoundary
```

Comment:
We should just use an `AllocateNode` then.

Reply:
Or I'll add a helper like `MakeStringNode` that takes a rope and a logical class.

Reply:
We probably should now this is not all hidden behind multiple truffle boundaries and so forth. Are you happy for that to be done in a separate PR along with @eregon's suggestion for changing pre_match, post_match, and to_s?

Reply:
We need to look at substring creation in this area anyway, because I'm not sure we're propagating taints correctly at the moment. I don't think we were before either.

Reply:
Future PR is fine. As long as there's a sensible method to search for. In this case, usages of `StringLayout#build` are good candidates.